File: nanocodex/agent/session_index.py
"""Session snapshots: a browsable, replayable history of past conversations.

The GUI shows a left-hand list of past conversations; clicking one replays its
FULL transcript. Two layers make that work:

* a small global INDEX at ``~/.nanocodex/sessions.jsonl`` — one line per
  conversation (a lightweight summary: title, counts, timestamps), newest-first
  for the list, and
* a per-conversation SNAPSHOT at ``~/.nanocodex/snapshots/<session_id>.json``
  holding the frozen full message list, so the detail view shows the real
  conversation — not a digest, and not the live ``session.jsonl`` (which
  ``--resume`` keeps appending to and compaction rewrites).

Keying is by **session_id**, NOT workspace: each time the GUI opens a project it
mints a new id, so re-opening or resuming the same folder keeps a SEPARATE
history entry instead of overwriting the previous one. Within one conversation
the same id is reused, so each turn rewrites (grows) that one snapshot in place.

Design split (same spirit as schedule.py / memory_store.py): SUMMARY EXTRACTION
and the STORE are pure over data (+ an injected "now"), so they unit-test fully
offline. Legacy workspace-keyed rows (the pre-snapshot format) still load — they
get a synthetic ``legacy:<workspace>`` id and simply have no snapshot to replay.
"""

from __future__ import annotations

import json
import uuid
from dataclasses import asdict, dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

DEFAULT_INDEX_PATH = Path.home() / ".nanocodex" / "sessions.jsonl"

# Cap the stored title/snippet so one giant paste can't bloat the index file.
_TITLE_MAX = 120
_SNIPPET_MAX = 200


def new_session_id() -> str:
    """Mint a fresh conversation id (the GUI calls this when it builds a loop)."""
    return uuid.uuid4().hex[:12]


def _now_iso() -> str:
    return datetime.now().isoformat(timespec="seconds")


def _first_text(content: "str | list[dict[str, Any]] | None") -> str:
    """Flatten a message ``content`` (str or block list) to plain text."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts: list[str] = []
        for block in content:
            if isinstance(block, dict) and block.get("type") == "text":
                parts.append(str(block.get("text", "")))
        return " ".join(p for p in parts if p)
    return ""


def _clip(text: str, limit: int) -> str:
    text = " ".join(text.split())  # collapse whitespace/newlines
    if len(text) <= limit:
        return text
    return text[: limit - 1].rstrip() + "…"


@dataclass
class SessionSummary:
    """A browsable digest of one conversation.

    Pure data — built by :func:`summarize`, persisted by :class:`SessionIndex`.
    ``session_id`` is the stable key (one entry per conversation; re-opening a
    project mints a new id and a new entry). The rest is a deterministic readout
    of the transcript for the directory list + detail view. ``has_snapshot``
    tells the GUI whether a full-transcript replay is available (legacy rows
    have none).
    """

    session_id: str
    workspace: str
    title: str = ""              # first user line, clipped — the list label
    snippet: str = ""            # last assistant line, clipped — a preview
    user_messages: int = 0
    assistant_messages: int = 0
    tool_calls: int = 0
    recent_tools: list[str] = field(default_factory=list)
    created_at: str = ""         # ISO timestamp of the first turn in this session
    updated_at: str = ""         # ISO timestamp of the last turn
    log_path: str = ""           # the workspace's session.jsonl, for reopening
    has_snapshot: bool = False   # whether a frozen full transcript exists

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def summarize(
    session_id: str,
    workspace: str,
    messages: list[dict[str, Any]],
    *,
    log_path: str = "",
    now_iso: str | None = None,
    created_at: str | None = None,
    has_snapshot: bool = False,
) -> SessionSummary:
    """Build a DETERMINISTIC summary of *messages* for one conversation.

    Zero-cost: title = first user message (clipped), snippet = last assistant
    message (clipped), plus counts and the most-recent tool names. No model
    call, no network — just a readout of the list. ``system`` messages are
    ignored (they're prompt scaffolding, not conversation).
    """
    title = ""
    snippet = ""
    users = assistants = tools = 0
    recent_tools: list[str] = []

    for m in messages:
        role = m.get("role")
        if role == "user":
            users += 1
            if not title:
                text = _first_text(m.get("content"))
                # Skip a compaction/summary marker injected as a user message.
                if text and not text.startswith("[Earlier conversation"):
                    title = _clip(text, _TITLE_MAX)
        elif role == "assistant":
            assistants += 1
            text = _first_text(m.get("content"))
            if text:
                snippet = _clip(text, _SNIPPET_MAX)
            for tc in m.get("tool_calls") or []:
                name = (tc.get("function") or {}).get("name")
                if name:
                    tools += 1
                    recent_tools.append(str(name))

    now = now_iso or _now_iso()
    return SessionSummary(
        session_id=session_id,
        workspace=workspace,
        title=title or "(no prompt yet)",
        snippet=snippet,
        user_messages=users,
        assistant_messages=assistants,
        tool_calls=tools,
        recent_tools=recent_tools[-8:],
        created_at=created_at or now,
        updated_at=now,
        log_path=log_path,
        has_snapshot=has_snapshot,
    )


def _redact_messages(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Strip base64 image data from a message list before freezing a snapshot.

    Mirrors session.py's log redaction: a full transcript may carry huge
    ``data:`` image URLs that would bloat the snapshot file (and aren't useful
    to replay as text). Replace each with a placeholder; everything else is kept
    verbatim so the conversation reads back faithfully.
    """
    out: list[dict[str, Any]] = []
    for msg in messages:
        content = msg.get("content")
        if not isinstance(content, list):
            out.append(msg)
            continue
        blocks: list[Any] = []
        changed = False
        for block in content:
            if (
                isinstance(block, dict)
                and block.get("type") == "image_url"
                and isinstance(block.get("image_url"), dict)
                and str(block["image_url"].get("url", "")).startswith("data:")
            ):
                blocks.append({"type": "text", "text": "[image omitted from snapshot]"})
                changed = True
            else:
                blocks.append(block)
        out.append({**msg, "content": blocks} if changed else msg)
    return out


class SessionIndex:
    """Load/save the global conversation directory as JSONL, keyed by session_id.

    Each conversation gets one row; :meth:`record` UPSERTs by ``session_id`` so a
    turn updates its own row, while a NEW conversation (new id) adds a separate
    entry — re-opening a project no longer overwrites its past history.
    :meth:`entries` returns them newest-activity-first for the directory list.

    Full transcripts are frozen as per-conversation snapshot files under
    ``<index dir>/snapshots/<session_id>.json`` so the detail view can replay the
    real conversation rather than a digest.
    """

    def __init__(self, path: Path | None = None) -> None:
        self.path = Path(path) if path is not None else DEFAULT_INDEX_PATH
        self.snapshots_dir = self.path.parent / "snapshots"
        self._by_id: dict[str, SessionSummary] = {}
        self._load()

    def _load(self) -> None:
        try:
            text = self.path.read_text(encoding="utf-8")
        except (OSError, ValueError):
            return
        for line in text.splitlines():
            line = line.strip()
            if not line:
                continue
            try:
                d = json.loads(line)
            except json.JSONDecodeError:
                continue  # tolerate a partially-written trailing line
            if not isinstance(d, dict):
                continue
            sid = d.get("session_id")
            ws = str(d.get("workspace", ""))
            # Legacy rows (pre-snapshot format) had no session_id, only a
            # workspace. Give them a stable synthetic id so they still list (they
            # simply have no snapshot to replay).
            if not sid:
                if not ws:
                    continue
                sid = f"legacy:{ws}"
            sid = str(sid)
            # Last write for an id wins (the file is append-style on rewrite).
            self._by_id[sid] = SessionSummary(
                session_id=sid,
                workspace=ws,
                title=str(d.get("title", "")),
                snippet=str(d.get("snippet", "")),
                user_messages=int(d.get("user_messages", 0) or 0),
                assistant_messages=int(d.get("assistant_messages", 0) or 0),
                tool_calls=int(d.get("tool_calls", 0) or 0),
                recent_tools=[str(t) for t in (d.get("recent_tools") or [])],
                created_at=str(d.get("created_at", "") or d.get("updated_at", "")),
                updated_at=str(d.get("updated_at", "")),
                log_path=str(d.get("log_path", "")),
                has_snapshot=bool(d.get("has_snapshot", False)),
            )

    def _save(self) -> None:
        """Rewrite the whole index file from the folded map (dedup on disk too)."""
        try:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            lines = [
                json.dumps(s.to_dict(), ensure_ascii=False)
                for s in self._sorted()
            ]
            self.path.write_text("\n".join(lines) + ("\n" if lines else ""),
                                 encoding="utf-8")
        except OSError:
            pass  # best-effort; a directory listing must never crash a turn

    def _sorted(self) -> list[SessionSummary]:
        """Newest activity first; blank timestamps sort last."""
        return sorted(
            self._by_id.values(),
            key=lambda s: s.updated_at or "",
            reverse=True,
        )

    # --- full-transcript snapshots ---------------------------------------

    def snapshot_path(self, session_id: str) -> Path:
        return self.snapshots_dir / f"{session_id}.json"

    def save_snapshot(self, session_id: str, messages: list[dict[str, Any]]) -> bool:
        """Freeze the full message list for one conversation. Best-effort.

        Returns True if the snapshot was written. Images are redacted to keep the
        file small; the snapshot is rewritten in full each turn (the transcript
        only grows), so the latest write always holds the whole conversation.
        """
        try:
            self.snapshots_dir.mkdir(parents=True, exist_ok=True)
            payload = {
                "session_id": session_id,
                "messages": _redact_messages(messages),
            }
            self.snapshot_path(session_id).write_text(
                json.dumps(payload, ensure_ascii=False), encoding="utf-8",
            )
            return True
        except OSError:
            return False

    def load_snapshot(self, session_id: str) -> list[dict[str, Any]] | None:
        """Read back a frozen transcript, or None if there is no snapshot."""
        try:
            text = self.snapshot_path(session_id).read_text(encoding="utf-8")
        except (OSError, ValueError):
            return None
        try:
            d = json.loads(text)
        except json.JSONDecodeError:
            return None
        msgs = d.get("messages") if isinstance(d, dict) else None
        return msgs if isinstance(msgs, list) else None

    # --- index records ---------------------------------------------------

    def record(self, summary: SessionSummary) -> None:
        """UPSERT one conversation's summary (by session_id) and persist."""
        self._by_id[summary.session_id] = summary
        self._save()

    def record_turn(
        self,
        session_id: str,
        workspace: str,
        messages: list[dict[str, Any]],
        *,
        log_path: str = "",
        now_iso: str | None = None,
    ) -> SessionSummary:
        """Freeze the transcript + upsert the row — the GUI's one call per turn.

        ``created_at`` is preserved from the conversation's first turn so the
        history keeps a stable "started at" even as later turns roll the
        ``updated_at`` forward.
        """
        prior = self._by_id.get(session_id)
        created_at = prior.created_at if prior else None
        saved = self.save_snapshot(session_id, messages)
        summary = summarize(
            session_id, workspace, messages,
            log_path=log_path, now_iso=now_iso,
            created_at=created_at, has_snapshot=saved,
        )
        self.record(summary)
        return summary

    def entries(self) -> list[SessionSummary]:
        """All conversation summaries, newest activity first (directory list)."""
        return self._sorted()

    def get(self, session_id: str) -> SessionSummary | None:
        return self._by_id.get(session_id)

File: rust/crates/ncx-core/src/session_index.rs
//! Browsable session history and frozen snapshots.
//!
//! The workspace JSONL log is for `--resume`; this global index is for a
//! human-facing directory of conversations. Each conversation has one summary
//! row keyed by a session id, plus a snapshot file with the full transcript.

use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

use serde_json::{json, Value};

use crate::session::{redact_image_data, Session};

const TITLE_MAX: usize = 120;
const SNIPPET_MAX: usize = 200;

static SESSION_SEQ: AtomicU64 = AtomicU64::new(0);

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SessionSummary {
    pub session_id: String,
    pub workspace: String,
    pub title: String,
    pub snippet: String,
    pub user_messages: usize,
    pub assistant_messages: usize,
    pub tool_calls: usize,
    pub recent_tools: Vec<String>,
    pub created_at: String,
    pub updated_at: String,
    pub log_path: String,
    pub has_snapshot: bool,
    pub archived: bool,
}

impl SessionSummary {
    fn to_value(&self) -> Value {
        json!({
            "session_id": self.session_id,
            "workspace": self.workspace,
            "title": self.title,
            "snippet": self.snippet,
            "user_messages": self.user_messages,
            "assistant_messages": self.assistant_messages,
            "tool_calls": self.tool_calls,
            "recent_tools": self.recent_tools,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "log_path": self.log_path,
            "has_snapshot": self.has_snapshot,
            "archived": self.archived,
        })
    }

    fn from_value(value: &Value) -> Option<Self> {
        let workspace = value
            .get("workspace")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let session_id = value
            .get("session_id")
            .and_then(|v| v.as_str())
            .map(str::to_string)
            .or_else(|| {
                if workspace.is_empty() {
                    None
                } else {
                    Some(format!("legacy:{workspace}"))
                }
            })?;
        Some(SessionSummary {
            session_id,
            workspace,
            title: string_field(value, "title"),
            snippet: string_field(value, "snippet"),
            user_messages: usize_field(value, "user_messages"),
            assistant_messages: usize_field(value, "assistant_messages"),
            tool_calls: usize_field(value, "tool_calls"),
            recent_tools: value
                .get("recent_tools")
                .and_then(|v| v.as_array())
                .map(|items| {
                    items
                        .iter()
                        .filter_map(|v| v.as_str().map(str::to_string))
                        .collect()
                })
                .unwrap_or_default(),
            created_at: string_field(value, "created_at"),
            updated_at: string_field(value, "updated_at"),
            log_path: string_field(value, "log_path"),
            has_snapshot: value
                .get("has_snapshot")
                .and_then(|v| v.as_bool())
                .unwrap_or(false),
            archived: value.get("archived").and_then(|v| v.as_bool()).unwrap_or(false),
        })
    }
}

pub struct SessionIndex {
    path: PathBuf,
    snapshots_dir: PathBuf,
    by_id: HashMap<String, SessionSummary>,
}

impl Default for SessionIndex {
    fn default() -> Self {
        Self::new(default_index_path())
    }
}

impl SessionIndex {
    pub fn new(path: PathBuf) -> Self {
        let snapshots_dir = path
            .parent()
            .map(|p| p.join("snapshots"))
            .unwrap_or_else(|| PathBuf::from("snapshots"));
        let mut index = SessionIndex {
            path,
            snapshots_dir,
            by_id: HashMap::new(),
        };
        index.load();
        index
    }

    pub fn entries(&self) -> Vec<SessionSummary> {
        let mut out = self.by_id.values().cloned().collect::<Vec<_>>();
        out.sort_by(|a, b| b.updated_at.cmp(&a.updated_at));
        out
    }

    pub fn get(&self, session_id: &str) -> Option<&SessionSummary> {
        self.by_id.get(session_id)
    }

    pub fn record(&mut self, summary: SessionSummary) {
        self.by_id.insert(summary.session_id.clone(), summary);
        self.save();
    }

    pub fn record_turn(
        &mut self,
        session_id: &str,
        workspace: &Path,
        session: &Session,
        log_path: &Path,
    ) -> SessionSummary {
        let prior_created = self.by_id.get(session_id).map(|s| s.created_at.clone());
        let prior_archived = self.by_id.get(session_id).map(|s| s.archived).unwrap_or(false);
        let saved = self.save_snapshot(session_id, session);
        let mut summary = summarize(
            session_id,
            &workspace.display().to_string(),
            &session.full_messages(),
            &log_path.display().to_string(),
            Some(now_stamp()),
            prior_created.as_deref(),
            saved,
        );
        summary.archived = prior_archived; // archiving survives new turns
        self.record(summary.clone());
        summary
    }

    /// Set a session's archived flag (persists). Returns false if unknown.
    pub fn set_archived(&mut self, session_id: &str, archived: bool) -> bool {
        match self.by_id.get_mut(session_id) {
            Some(s) => {
                s.archived = archived;
                self.save();
                true
            }
            None => false,
        }
    }

    pub fn snapshot_path(&self, session_id: &str) -> PathBuf {
        self.snapshots_dir
            .join(format!("{}.json", safe_file_stem(session_id)))
    }

    pub fn save_snapshot(&self, session_id: &str, session: &Session) -> bool {
        let payload = json!({
            "session_id": session_id,
            "messages": redact_messages(&session.full_messages(), "[image omitted from snapshot]"),
        });
        if fs::create_dir_all(&self.snapshots_dir).is_err() {
            return false;
        }
        serde_json::to_string(&payload)
            .ok()
            .and_then(|text| fs::write(self.snapshot_path(session_id), text).ok())
            .is_some()
    }

    pub fn load_snapshot(&self, session_id: &str) -> Option<Vec<Value>> {
        let text = fs::read_to_string(self.snapshot_path(session_id)).ok()?;
        let value = serde_json::from_str::<Value>(&text).ok()?;
        value.get("messages")?.as_array().cloned()
    }

    fn load(&mut self) {
        let Ok(text) = fs::read_to_string(&self.path) else {
            return;
        };
        for line in text.lines() {
            let Ok(value) = serde_json::from_str::<Value>(line.trim()) else {
                continue;
            };
            let Some(summary) = SessionSummary::from_value(&value) else {
                continue;
            };
            self.by_id.insert(summary.session_id.clone(), summary);
        }
    }

    fn save(&self) {
        if let Some(parent) = self.path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let lines = self
            .entries()
            .iter()
            .filter_map(|s| serde_json::to_string(&s.to_value()).ok())
            .collect::<Vec<_>>();
        let text = if lines.is_empty() {
            String::new()
        } else {
            format!("{}\n", lines.join("\n"))
        };
        let _ = fs::write(&self.path, text);
    }
}

pub fn new_session_id() -> String {
    let n = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let seq = SESSION_SEQ.fetch_add(1, Ordering::Relaxed);
    format!("{n:x}{:x}{seq:x}", std::process::id())
}

pub fn default_index_path() -> PathBuf {
    let home = std::env::var_os("USERPROFILE")
        .or_else(|| std::env::var_os("HOME"))
        .map(PathBuf::from)
        .unwrap_or_default();
    home.join(".nanocodex").join("sessions.jsonl")
}

pub fn summarize(
    session_id: &str,
    workspace: &str,
    messages: &[Value],
    log_path: &str,
    now: Option<String>,
    created_at: Option<&str>,
    has_snapshot: bool,
) -> SessionSummary {
    let mut title = String::new();
    let mut snippet = String::new();
    let mut user_messages = 0;
    let mut assistant_messages = 0;
    let mut tool_calls = 0;
    let mut recent_tools = Vec::new();

    for msg in messages {
        match msg.get("role").and_then(|v| v.as_str()) {
            Some("user") => {
                user_messages += 1;
                if title.is_empty() {
                    let text = first_text(msg.get("content"));
                    if !text.is_empty() && !text.starts_with("[Earlier conversation") {
                        title = clip(&text, TITLE_MAX);
                    }
                }
            }
            Some("assistant") => {
                assistant_messages += 1;
                let text = first_text(msg.get("content"));
                if !text.is_empty() {
                    snippet = clip(&text, SNIPPET_MAX);
                }
                if let Some(calls) = msg.get("tool_calls").and_then(|v| v.as_array()) {
                    tool_calls += calls.len();
                    for call in calls {
                        if let Some(name) = call
                            .get("function")
                            .and_then(|f| f.get("name"))
                            .and_then(|v| v.as_str())
                        {
                            recent_tools.push(name.to_string());
                        }
                    }
                }
            }
            _ => {}
        }
    }

    if recent_tools.len() > 8 {
        recent_tools = recent_tools[recent_tools.len() - 8..].to_vec();
    }
    let now = now.unwrap_or_else(now_stamp);
    SessionSummary {
        session_id: session_id.to_string(),
        workspace: workspace.to_string(),
        title: if title.is_empty() {
            "(no prompt yet)".into()
        } else {
            title
        },
        snippet,
        user_messages,
        assistant_messages,
        tool_calls,
        recent_tools,
        created_at: created_at.unwrap_or(&now).to_string(),
        updated_at: now,
        log_path: log_path.to_string(),
        has_snapshot,
        archived: false,
    }
}

fn first_text(content: Option<&Value>) -> String {
    match content {
        Some(Value::String(s)) => s.clone(),
        Some(Value::Array(blocks)) => blocks
            .iter()
            .filter(|b| b.get("type").and_then(|v| v.as_str()) == Some("text"))
            .filter_map(|b| b.get("text").and_then(|v| v.as_str()))
            .filter(|s| !s.is_empty())
            .collect::<Vec<_>>()
            .join(" "),
        _ => String::new(),
    }
}

fn clip(text: &str, limit: usize) -> String {
    let collapsed = text.split_whitespace().collect::<Vec<_>>().join(" ");
    if collapsed.chars().count() <= limit {
        return collapsed;
    }
    let take = limit.saturating_sub(3);
    format!("{}...", collapsed.chars().take(take).collect::<String>())
}

fn redact_messages(messages: &[Value], placeholder: &str) -> Vec<Value> {
    messages
        .iter()
        .map(|msg| redact_image_data(msg, placeholder))
        .collect()
}

fn string_field(value: &Value, key: &str) -> String {
    value
        .get(key)
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string()
}

fn usize_field(value: &Value, key: &str) -> usize {
    value
        .get(key)
        .and_then(|v| v.as_u64())
        .and_then(|v| usize::try_from(v).ok())
        .unwrap_or(0)
}

fn safe_file_stem(input: &str) -> String {
    input
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '-' || c == '_' {
                c
            } else {
                '_'
            }
        })
        .collect()
}

fn now_stamp() -> String {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| format!("{:013}", d.as_millis()))
        .unwrap_or_else(|_| "0000000000000".into())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tmp_path(name: &str) -> PathBuf {
        std::env::temp_dir().join(format!("ncx_session_index_{name}_{}", now_stamp()))
    }

    fn msgs() -> Vec<Value> {
        vec![
            json!({"role": "system", "content": "sys"}),
            json!({"role": "user", "content": "fix login"}),
            json!({"role": "assistant", "content": "looking", "tool_calls": [
                {"id": "1", "type": "function", "function": {"name": "read_file", "arguments": "{}"}}
            ]}),
            json!({"role": "tool", "tool_call_id": "1", "name": "read_file", "content": "..."}),
            json!({"role": "assistant", "content": "fixed"}),
        ]
    }

    #[test]
    fn summarize_pulls_title_snippet_counts_and_tools() {
        let s = summarize(
            "sid",
            "/proj",
            &msgs(),
            "/proj/.nanocodex/session.jsonl",
            Some("2026-06-01T10:00:00".into()),
            None,
            true,
        );
        assert_eq!(s.title, "fix login");
        assert_eq!(s.snippet, "fixed");
        assert_eq!(s.user_messages, 1);
        assert_eq!(s.assistant_messages, 2);
        assert_eq!(s.tool_calls, 1);
        assert_eq!(s.recent_tools, vec!["read_file"]);
        assert_eq!(s.created_at, "2026-06-01T10:00:00");
        assert!(s.has_snapshot);
    }

    #[test]
    fn index_upserts_and_sorts_newest_first() {
        let path = tmp_path("sort").join("sessions.jsonl");
        let mut idx = SessionIndex::new(path);
        idx.record(summarize(
            "old",
            "/p",
            &msgs(),
            "",
            Some("2026-06-01T09:00:00".into()),
            None,
            false,
        ));
        idx.record(summarize(
            "new",
            "/p",
            &msgs(),
            "",
            Some("2026-06-01T11:00:00".into()),
            None,
            false,
        ));
        idx.record(summarize(
            "old",
            "/p",
            &msgs(),
            "",
            Some("2026-06-01T12:00:00".into()),
            Some("2026-06-01T09:00:00"),
            false,
        ));

        let entries = idx.entries();
        assert_eq!(entries.len(), 2);
        assert_eq!(entries[0].session_id, "old");
        assert_eq!(entries[0].created_at, "2026-06-01T09:00:00");
        assert_eq!(entries[1].session_id, "new");
    }

    #[test]
    fn persists_and_loads_legacy_rows() {
        let dir = tmp_path("legacy");
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("sessions.jsonl");
        std::fs::write(
            &path,
            "{\"workspace\":\"/old\",\"title\":\"legacy\",\"updated_at\":\"2026\"}\nnot json\n",
        )
        .unwrap();

        let idx = SessionIndex::new(path);
        let entries = idx.entries();
        assert_eq!(entries.len(), 1);
        assert_eq!(entries[0].session_id, "legacy:/old");
        assert_eq!(entries[0].title, "legacy");
    }

    #[test]
    fn snapshot_round_trip_redacts_image_data() {
        let dir = tmp_path("snapshot");
        let mut idx = SessionIndex::new(dir.join("sessions.jsonl"));
        let mut session = Session::new("sys");
        session.add_user(json!([
            {"type": "text", "text": "describe"},
            {"type": "image_url", "image_url": {"url": "data:image/png;base64,AAAA"}}
        ]));

        idx.record_turn(
            "sid",
            Path::new("/p"),
            &session,
            Path::new("/p/.nanocodex/session.jsonl"),
        );
        let loaded = idx.load_snapshot("sid").unwrap();
        let text = serde_json::to_string(&loaded).unwrap();
        assert!(text.contains("[image omitted from snapshot]"));
        assert!(!text.contains("data:image"));
        assert!(idx.get("sid").unwrap().has_snapshot);
    }

    #[test]
    fn session_ids_are_unique() {
        assert_ne!(new_session_id(), new_session_id());
    }
}
